### DISCUSSION TITLE
Support for Sustain Pedal / Midi CC64

### DISCUSSION THREAD

**Discussion by Psirus** [SynthstromAudible/DelugeFirmware#264] [Ideas]
I'd love to see support for a sustain pedal, i.e. Midi CC64. [There is also interest expressed in the forums.](https://forums.synthstrom.com/discussion/4111/support-for-external-midi-sustain-in-4-0)

I'm willing to try implementing it myself, as soon as the cross-platform build scripts are available. Maybe anyone has any pointers on how this is usually done.

This is just to gauge interest, and maybe collect some additional requirements/wishes, so let me know what you think.

**Comment by dctucker** [comment]
I would get a lot of use out of this feature being implemented. I have some criteria to add that I think would make it even more useful:

- When a MIDI track has learned note on events for a given channel, it should also recognize and forward CC64 as sustain, if not by default then possibly by listening for the presence of CC64 when holding learn and pressing an audition pad.

- Sustain also applies to keep arpeggiators running

**Comment by manysounds** [comment]
It is ridiculous that this isn’t the default behavior however cc64 needs to be repatchable for instances where a user may want their sustain pedal tied to another function, like how some organ synths elsewhere use 64 for Leslie speed.
Eventually I would love to see a menu-list that shows what CCs are assigned to what.
Furthermore, being able to assign specific CCs intentionally without relying on *catching* them would be EXTREMELY useful for MPE controllers since they will often send -pressure- and -slide- simultaneously and linking those often goes awry.

**Comment by weavermedia** [comment]
It would be cool if we could have a sustain 'chord' mode too, similar to the Midihub software from Blokas:

**Sustain mode**
- All - Sustain all notes
- Chord - Add sustained notes while at least one note is held down, once all keys are up and a new note is pressed, old notes will be turned off and newly pressed ones will be sustained.

**Comment by o010b** [comment]
This feels like a topic that might expand rapidly into an entire project and CC overhaul that may or may not happen in unknown future time.

I'd like to suggest a fix until then that sustain pedal acts as a real time modifier at playback/record time, so that sustain pedal simply holds release of notes until the sustain pedal is released.

It would work satisfactory, it would not create complications with the user interface, it would make sense for the majority of use cases with the Deluge. And it's better than nothing. 

Edit: Oh, wait. This is how it works in MIDI mode, but not in synth mode for some reason?

**Comment by Maly-Lemire** [comment]
It seems to work in midi mode but it does not support the arpegiator.

**Comment by anthony-amerman** [comment]
i would love this.

**Comment by robenimmersatt** [comment]
Would love this for the DX7 EP sounds.

**Comment by robenimmersatt** [comment]
I would expect that while the sustain pedal is pressed, no note off messages would be passed thru. Is that so hard to implement? 
Might also have a look at the code, though this is new territory for me (good thing we have llms these days).

**Reply by weavermedia** [reply]
You'd just have to store the note numbers so you can send all the NoteOff messages when the pedal is released.

**Comment by vlork** [comment]
I'd like to suggest a sustain pedal pre-record transform so that notes played with sustain become full length in the clips as if held down for the entire duration. Both in MIDI and Synth Clips. Because the Deluge has no immediate way to display pedal on/off and the results are identical for 99.9% percent of users.

**Comment by seangoodvibes** [comment]
> I'd like to suggest a sustain pedal pre-record transform so that notes played with sustain become full length in the clips as if held down for the entire duration. Both in MIDI and Synth Clips. Because the Deluge has no immediate way to display pedal on/off and the results are identical for 99.9% percent of users.

I think the idea was to record pedal automation so you can see the pedal on / off in automation view

There's a PR here where someone was working on it:

https://github.com/SynthstromAudible/DelugeFirmware/pull/3404

**Comment by seangoodvibes** [comment]
> I would expect that while the sustain pedal is pressed, no note off messages would be passed thru. Is that so hard to implement? 
> 
> Might also have a look at the code, though this is new territory for me (good thing we have llms these days).
> 
> 

See: https://github.com/SynthstromAudible/DelugeFirmware/pull/3404
